diff -r -u treeCl-old/setup.py treeCl-new/setup.py
--- treeCl-old/setup.py	2019-06-05 14:29:18.046251971 -0700
+++ treeCl-new/setup.py	2019-06-05 14:28:51.332426008 -0700
@@ -79,7 +79,7 @@
           # TODO why were these commented out?? ask author if they work
           # 'bin/simulator',
           # 'bin/collapse',
-          # 'bin/treeCl',
+          'bin/treeCl',
           # 'bin/seqconvert',
           # 'bin/bootstrap',
           # 'bin/npbs.py',
