diff -ruN sonicparanoid-1.2.3/sonicparanoid/mcl_c.pyx sonicparanoid-patched/sonicparanoid/mcl_c.pyx
--- sonicparanoid-1.2.3/sonicparanoid/mcl_c.pyx	2019-06-07 00:06:32.000000000 +0000
+++ sonicparanoid-patched/sonicparanoid/mcl_c.pyx	2019-06-29 00:49:13.742000000 +0000
@@ -90,7 +90,7 @@
 
   # make sure MCL is installed
   # check if mcl has been installed
-  mclPath: str = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'mcl_package/bin/mcl')
+  mclPath = '@mcl@/bin/mcl'
   if os.path.isfile(mclPath):
     if debug:
       print("MCL is installed at {:s}".format(mclPath))
diff -ruN sonicparanoid-1.2.3/sonicparanoid/workers.py sonicparanoid-patched/sonicparanoid/workers.py
--- sonicparanoid-1.2.3/sonicparanoid/workers.py	2019-06-07 00:35:56.000000000 +0000
+++ sonicparanoid-patched/sonicparanoid/workers.py	2019-06-29 00:50:29.474000000 +0000
@@ -255,9 +255,7 @@
 def get_mmseqs_path():
     """Return the directory in which MMseqs2 binaries are stored."""
     #import platform
-    mmseqsPath = None
-    pySrcDir = os.path.dirname(os.path.abspath(__file__))
-    mmseqsPath = os.path.join(pySrcDir, 'bin/mmseqs')
+    mmseqsPath = '@mmseqs2@/bin/mmseqs'
     if not os.path.isfile(mmseqsPath):
         sys.stderr.write('\nERROR: mmseqs2 was not found, please install it and execute setup_sonicparanoid.py.')
     # return the path
