diff -ruN sonicparanoid-old/sonicparanoid/sonic_paranoid.py sonicparanoid-new/sonicparanoid/sonic_paranoid.py
--- sonicparanoid-old/sonicparanoid/sonic_paranoid.py	2019-05-12 18:41:32.000000000 -0700
+++ sonicparanoid-new/sonicparanoid/sonic_paranoid.py	2019-06-06 17:05:29.985616881 -0700
@@ -631,7 +631,7 @@
     #start setting the needed variables
     debug = args.debug
     # check MMseqs2 installation
-    check_mmseqs_installation(root, debug=debug)
+    # check_mmseqs_installation(root, debug=debug)
     # check MCL installation
     check_mcl_installation(root, debug=debug)
     inDir = None
