diff -ruN src-old/config/config.ini.default src-new/config/config.ini.default
--- busco/config/config.ini.default	2019-06-10 11:07:39.375403677 -0700
+++ busco/config/config.ini.default	2019-06-10 11:09:54.852525573 -0700
@@ -45,32 +45,32 @@
 
 [tblastn]
 # path to tblastn
-path = /usr/bin/
+path = @myBlast@/bin/
 
 [makeblastdb]
 # path to makeblastdb
-path = /usr/bin/
+path = @myBlast@/bin/
 
 [augustus]
 # path to augustus
-path = /home/osboxes/BUSCOVM/augustus/augustus-3.2.2/bin/
+path = @augustus@/bin/
 
 [etraining]
 # path to augustus etraining
-path = /home/osboxes/BUSCOVM/augustus/augustus-3.2.2/bin/
+path = @augustus@/bin/
 
 # path to augustus perl scripts, redeclare it for each new script
 [gff2gbSmallDNA.pl]
-path = /home/osboxes/BUSCOVM/augustus/augustus-3.2.2/scripts/
+path = @augustus@/scripts/
 [new_species.pl]
-path = /home/osboxes/BUSCOVM/augustus/augustus-3.2.2/scripts/
+path = @augustus@/scripts/
 [optimize_augustus.pl]
-path = /home/osboxes/BUSCOVM/augustus/augustus-3.2.2/scripts/
+path = @augustus@/scripts/
 
 [hmmsearch]
 # path to HMMsearch executable
-path = /home/osboxes/BUSCOVM/hmmer/hmmer-3.1b2-linux-intel-ia32/binaries/
+path = @hmmer@/bin/
 
 [Rscript]
 # path to Rscript, if you wish to use the plot tool
-path = /usr/bin/
+path = @myR@/bin/
