--- blastdbget/blastdbget.py	2023-07-23 11:57:22.845000000 -0700
+++ blastdbget-no-validate-dbs/blastdbget.py	2023-07-23 11:55:01.937000000 -0700
@@ -21,7 +21,7 @@
 import urllib
 
 from contextlib import contextmanager
-from sh import blastdbcheck, ErrorReturnCode
+from sh import ErrorReturnCode
 from tempfile import mkdtemp
 from threading import Thread
 from Queue import Queue
@@ -301,7 +301,7 @@
         show_available(remote_files)
 
     if not "taxdb" in database:
-        database + ("taxdb",)
+        database += ["taxdb"]
 
     if output is None:
         output = safe_makedir(os.path.abspath('.'))
@@ -325,13 +325,14 @@
     q.join()
 
     # validate the database
-    status = validate_dbs(output, database)
-    for db, success in status.iteritems():
-        if success:
-            log.info("%s created successfully" % db)
-            # maybe delete the tgz and md5 files
-        else:
-            log.warn("%s failed validation" % db)
+    # (removed because blastdbcheck seems to be broken)
+    # status = validate_dbs(output, database)
+    # for db, success in status.iteritems():
+    #     if success:
+    #         log.info("%s created successfully" % db)
+    #         # maybe delete the tgz and md5 files
+    #     else:
+    #         log.warn("%s failed validation" % db)
     log.info("Process complete.")
 
 
